### PR TITLE
R21.0 xmlhttp fixes

### PR DESCRIPTION
**Pull request by colman-hartinger**
Fix so that XMLHttpRequest has getRequestHeader function when running in browser. Fixed by changing function definition to add headers to headers object then defined get function to retrieve the headers. Also added to the ts definition using a global declare for the file.

### CODE REVIEW

File: src/api.ts [hunk 1]
```diff
@@ -11,12 +11,42 @@ import {
   Token,
 } from './types';
 
-let XMLHttpRequest;
 
 if (typeof window === 'undefined') {
-  XMLHttpRequest = require('xmlhttprequest').XMLHttpRequest;
+  var XMLHttpRequest = require('xmlhttprequest').XMLHttpRequest;
```
Comment (colman-hartinger) [inline src/api.ts:16]: had to use var here because let/const operated in the wrong scope

File: src/api.ts [hunk 2]
```diff
@@ -11,12 +11,42 @@ import {
...
 }
 
+declare global {
+  interface XMLHttpRequest {
+    headers: any
+    wrappedSetRequestHeader(header: string, value: string): void
+    getRequestHeader(type: any):string
```
Comment (bradymadams) [inline src/api.ts]: Space before string.

**Review by bradymadams — COMMENTED** [review]
Looks good, but can you explain why wrapping `setRequestHeader` is necessary? And how doing so will work in both browser and node?

File: src/api.ts (hunk 2 above)
Comment (bradymadams) [reply src/api.ts]: Nevermind, looks like you got to it already.

**Comment by colman-hartinger** [comment]
> Looks good, but can you explain why wrapping `setRequestHeader` is necessary? And how doing so will work in both browser and node?

The wrapper for `setRequestHeader` is necessary for creating the headers object in  `XMLHttpRequest` so that `getRequestHeader` can be used. With how the import is structured when running in node, the wrapper gets ignored and node just uses the default `XMLHttpRequest.getRequestHeader` function when running in the browser where `getRequestHeader` doesn't exist that function is now defined to return a key - value from the headers object created from the `setRequestHeader` wrapper

File: src/api.ts [hunk 3]
```diff
@@ -169,7 +164,8 @@ export class API {
       xhttp.send();
     } else if (data instanceof Buffer) {
       xhttp.setRequestHeader('Content-Type', 'model/3mf');
-      if (xhttp.getRequestHeader(EncodingTypes.content) == EncodingValues.gzip) {
+      if (encoding && encoding.value == EncodingValues.gzip) {
+        console.log('this. works')
```
Comment (bradymadams) [inline src/api.ts:168]: leftover debug.

**Review by bradymadams — CHANGES_REQUESTED** [review]
Can you explain all of the changes to how the `version` is used? Are they necessary for this PR? I thought I had it all setup correctly. And the `verifyVersion` function wouldn't be sending the client version to `success` anymore. Just the version sent back from the server via `Accept-Version`.

**Comment by colman-hartinger** [comment]
> Can you explain all of the changes to how the `version` is used? Are they necessary for this PR? I thought I had it all setup correctly. And the `verifyVersion` function wouldn't be sending the client version to `success` anymore. Just the version sent back from the server via `Accept-Version`.

Ya, I'm not sure how it made it through my last round of testing but the way I was defining the `static get version` was causing an infinite loop so I decided to just use the version object on API instead of the static get. I just changed things around to keep the original structure by using static get and removing version from API so the recursion won't happen

File: src/api.ts [hunk 4]
```diff
@@ -205,7 +202,7 @@ export class API {
       // how can we make this less restrictive?
       const compatible = (sv_maj === cv_maj && sv_min === cv_min);
 
-      success(compatible, API.version, this.version);
+      success(compatible, API.version, API.version);
```
Comment (bradymadams) [inline src/api.ts:205]: I think this change and the change on line 192 need to be reverted to use the version declared in this file.

**Review by bradymadams — CHANGES_REQUESTED** [review]

File: src/api.ts (hunk 4 above)
Comment (colman-hartinger) [reply src/api.ts]: your right, I should of caught that. Fixed now

**Review by bradymadams — APPROVED** [review]

**Pull request merged**